[DictionarySorter] Sorting Algorithm Optimized for Dictionary Encoding

CaiusDai opened the pull request:
# What is included in this PR
- A sorting algorithm optimized for Dictionary Encoding
- A mini benchmark file that compares the sorting speed with Arrow's Sorting Algorithm.

# Some important context
- Only support for single column sorting for now ( key column can only be 1, while the payload doesn't matter)
- Only works when all column chunks of the key column is encoded by Dictionary Encoding and no encoding fallback happens (i.e, when dictionary encoding switches to plain encoding if the page is full)

# Rest work
- Correctness checking code currently has a bug, need fix
- Better Apply code
- More comprehensive benchmark results are needed ( Data generator )

CaiusDai committed:
[Sort][Dict]Formulate&Init

CaiusDai committed:
[Bench]Modify bench code

CaiusDai committed:
[Fix]Minor Update

howarlii commented on benchmark/sortproto/parquet_sorter.cc:
```diff
@@ -0,0 +1,550 @@
+  // Merge the results from different streams
+  auto merge_result = merge_streams(streams);
+  // Construct a Index Mapping based on counting results.
+  auto offset_map = get_offset_mapping(std::move(merge_result));
```
我猜测CountBaseSort很大一部分时间都花费在了这个unordered_map上面，因为所有的元素都需要在这个offset_map上进行find，这将耗费大量的时间。
这里可以改成直接使用内存寻址的方式，考虑到sorted result中每一个元素都存在唯一的`<row_group, dict_pos>`二元组，可以使用这个开一个二维array来存储offset，而不是直接暴力使用 unordered_map

CaiusDai replied:
我先查一下profiling的结果吧。主要是merged sorted result中同个元素可能存在多个<RowGroup,DictPos>的映射（ 不同RowGroup可能存在同个Dictionary Element）， 要想用2D数组来记录offset可能需要多次查找并寻址Update。

howarlii replied:
我们不用特别考虑元素重复的情况呀，相同元素谁在前面谁在后面不重要，只要最后答案是有序的即可。换句话说，把dictionary排好序之后，具体元素值是多少就不重要了，这不会对答案有影响，dictionary中相同元素的值按照不同的值的方式处理 最后答案也是对的

CaiusDai committed:
[Fix] Fix bug, add bench tools

Add some utility functions and correctness checking code. Fixed bugs
relating to file writing and segmentation fault.

CaiusDai committed:
Improve CountBase
